File: src/caco/tui/screens/cache.py
"""Cache management screen."""

from pathlib import Path

from textual.app import ComposeResult
from textual.binding import Binding
from textual.screen import Screen
from textual.widgets import DataTable, Footer, Static

from caco import db


def _format_size(size_bytes: int) -> str:
    """Format bytes into human-readable size."""
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024 * 1024:
        return f"{size_bytes / 1024:.1f} KB"
    elif size_bytes < 1024 * 1024 * 1024:
        return f"{size_bytes / (1024 * 1024):.1f} MB"
    else:
        return f"{size_bytes / (1024 * 1024 * 1024):.2f} GB"


class CacheScreen(Screen):
    """Cache management screen showing cached WAD files."""

    BINDINGS = [
        Binding("q", "back", "Back", show=True),
        Binding("escape", "back", "Back", show=False),
        Binding("d", "clear_selected", "Clear", show=True),
        Binding("D", "clear_all", "Clear All", show=True, key_display="D"),
        Binding("j", "cursor_down", "Down", show=False),
        Binding("k", "cursor_up", "Up", show=False),
    ]

    DEFAULT_CSS = """
    CacheScreen {
        padding: 1 2;
    }

    CacheScreen #cache-header {
        text-style: bold;
        margin-bottom: 1;
    }

    CacheScreen #cache-table {
        height: 1fr;
    }

    CacheScreen #cache-status {
        dock: bottom;
        height: 1;
        padding: 0 1;
        color: $text-muted;
    }
    """

    def compose(self) -> ComposeResult:
        yield Static("[bold]Cache Management[/bold]", id="cache-header")
        yield DataTable(id="cache-table")
        yield Static("", id="cache-status")
        yield Footer()

    def on_mount(self) -> None:
        """Set up table and load cached WADs."""
        table = self.query_one("#cache-table", DataTable)
        table.cursor_type = "row"
        table.zebra_stripes = True
        table.add_column("ID", key="id", width=6)
        table.add_column("Title", key="title", width=30)
        table.add_column("Path", key="path", width=40)
        table.add_column("Size", key="size", width=10)
        self._load_cache()

    def _load_cache(self) -> None:
        """Load cached WADs into table."""
        table = self.query_one("#cache-table", DataTable)
        table.clear()

        self._cached = db.get_cached_wads()
        total_size = 0

        for wad in self._cached:
            path = Path(wad["cached_path"]) if wad.get("cached_path") else None
            if path and path.exists():
                size = path.stat().st_size
                total_size += size
                size_str = _format_size(size)
            else:
                size_str = "missing"

            table.add_row(
                str(wad["id"]),
                wad["title"],
                wad.get("cached_path", "-"),
                size_str,
                key=str(wad["id"]),
            )

        status = self.query_one("#cache-status", Static)
        status.update(f"{len(self._cached)} cached files | Total: {_format_size(total_size)} | d=Clear  D=Clear All")

    def action_clear_selected(self) -> None:
        """Clear cache for the selected WAD."""
        table = self.query_one("#cache-table", DataTable)
        if table.cursor_row is None or table.cursor_row >= len(self._cached):
            self.notify("No WAD selected", severity="warning")
            return

        wad = self._cached[table.cursor_row]
        wad_id = wad["id"]

        # Delete the file
        path = Path(wad["cached_path"]) if wad.get("cached_path") else None
        if path and path.exists():
            path.unlink()

        db.clear_cached_path(wad_id)
        self.notify(f"Cleared cache for {wad['title']}")
        self._load_cache()

    def action_clear_all(self) -> None:
        """Clear all cached WAD files."""
        for wad in self._cached:
            path = Path(wad["cached_path"]) if wad.get("cached_path") else None
            if path and path.exists():
                path.unlink()

        count = db.clear_all_cached_paths()
        self.notify(f"Cleared {count} cached files")
        self._load_cache()

    def action_cursor_down(self) -> None:
        """Move cursor down."""
        table = self.query_one("#cache-table", DataTable)
        if table.cursor_row is not None and table.cursor_row < len(self._cached) - 1:
            table.move_cursor(row=table.cursor_row + 1)

    def action_cursor_up(self) -> None:
        """Move cursor up."""
        table = self.query_one("#cache-table", DataTable)
        if table.cursor_row is not None and table.cursor_row > 0:
            table.move_cursor(row=table.cursor_row - 1)

    def action_back(self) -> None:
        """Go back."""
        self.app.pop_screen()

File: crates/caco-tui/src/screens/cache.rs
use std::fs;

use crate::widgets::table_nav::{table_nav_next, table_nav_prev};
use caco_core::db::sessions::{clear_all_cached_paths, clear_cached_path, get_cached_wads};
use caco_core::utils::format_size;
use crossterm::event::{KeyCode, KeyEvent};
use ratatui::Frame;
use ratatui::layout::{Constraint, Layout, Rect};
use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::{Block, BorderType, Borders, Cell, Paragraph, Row, Table, TableState};
use rusqlite::Connection;

use crate::message::{AppMessage, ScreenResult, Severity};
use crate::screens::Screen;
use crate::theme;

/// Cache entry with filesystem info.
struct CacheEntry {
    wad_id: i64,
    title: String,
    path: String,
    size: Option<u64>,
}

/// Cache management screen.
pub struct CacheScreen {
    entries: Vec<CacheEntry>,
    table_state: TableState,
    total_size: u64,
}

impl CacheScreen {
    pub fn new(conn: &Connection) -> Self {
        let mut screen = Self {
            entries: Vec::new(),
            table_state: TableState::default(),
            total_size: 0,
        };
        screen.load(conn);
        screen
    }

    fn load(&mut self, conn: &Connection) {
        let wads = get_cached_wads(conn).unwrap_or_default();
        self.total_size = 0;
        self.entries = wads
            .into_iter()
            .filter_map(|w| {
                let path = w.cached_path?;
                let size = fs::metadata(&path).ok().map(|m| m.len());
                if let Some(s) = size {
                    self.total_size += s;
                }
                Some(CacheEntry {
                    wad_id: w.id,
                    title: w.title,
                    path,
                    size,
                })
            })
            .collect();

        if !self.entries.is_empty() {
            let current = self.table_state.selected().unwrap_or(0);
            if current >= self.entries.len() {
                self.table_state.select(Some(self.entries.len() - 1));
            } else {
                self.table_state.select(Some(current));
            }
        } else {
            self.table_state.select(None);
        }
    }
}

impl Screen for CacheScreen {
    fn render(&mut self, frame: &mut Frame, area: Rect, _conn: &Connection) {
        let block = Block::default()
            .borders(Borders::ALL)
            .border_type(BorderType::Rounded)
            .border_style(theme::border_style())
            .title(format!(
                " Cache — {} files, {} ",
                self.entries.len(),
                format_size(self.total_size)
            ));
        let inner = block.inner(area);
        frame.render_widget(block, area);

        let layout = Layout::vertical([Constraint::Min(1), Constraint::Length(1)]).split(inner);

        if self.entries.is_empty() {
            frame.render_widget(
                Paragraph::new("Cache is empty").style(theme::dim_style()),
                layout[0],
            );
        } else {
            let header = Row::new(vec![
                Cell::from("ID").style(Style::default().add_modifier(Modifier::BOLD)),
                Cell::from("Title").style(Style::default().add_modifier(Modifier::BOLD)),
                Cell::from("Path").style(Style::default().add_modifier(Modifier::BOLD)),
                Cell::from("Size").style(Style::default().add_modifier(Modifier::BOLD)),
            ])
            .height(1);

            let rows: Vec<Row> = self
                .entries
                .iter()
                .map(|e| {
                    let size_str = match e.size {
                        Some(s) => format_size(s),
                        None => "missing".to_string(),
                    };
                    let size_style = if e.size.is_none() {
                        Style::default().fg(Color::Red)
                    } else {
                        Style::default()
                    };
                    Row::new(vec![
                        Cell::from(e.wad_id.to_string()),
                        Cell::from(e.title.clone()),
                        Cell::from(e.path.clone()),
                        Cell::from(size_str).style(size_style),
                    ])
                })
                .collect();

            let widths = [
                Constraint::Length(5),
                Constraint::Min(20),
                Constraint::Min(30),
                Constraint::Length(10),
            ];

            let table = Table::new(rows, widths)
                .header(header)
                .row_highlight_style(theme::highlight_style());
            frame.render_stateful_widget(table, layout[0], &mut self.table_state);
        }

        // Key hints
        let hints = Line::from(vec![
            Span::styled("q/Esc", theme::key_style()),
            Span::styled(" back  ", theme::desc_style()),
            Span::styled("d", theme::key_style()),
            Span::styled(" delete  ", theme::desc_style()),
            Span::styled("D", theme::key_style()),
            Span::styled(" delete all  ", theme::desc_style()),
            Span::styled("j/k", theme::key_style()),
            Span::styled(" navigate", theme::desc_style()),
        ]);
        frame.render_widget(Paragraph::new(hints), layout[1]);
    }

    fn handle_key(&mut self, key: KeyEvent, conn: &Connection) -> Option<AppMessage> {
        match key.code {
            KeyCode::Char('q') | KeyCode::Esc => {
                Some(AppMessage::PopScreen(ScreenResult::Cancelled))
            }
            KeyCode::Char('j') | KeyCode::Down => {
                table_nav_next(&mut self.table_state, self.entries.len());
                None
            }
            KeyCode::Char('k') | KeyCode::Up => {
                table_nav_prev(&mut self.table_state, self.entries.len());
                None
            }
            KeyCode::Char('d') => {
                // Delete selected cached file
                if let Some(idx) = self.table_state.selected() {
                    if let Some(entry) = self.entries.get(idx) {
                        let _ = fs::remove_file(&entry.path);
                        let _ = clear_cached_path(conn, entry.wad_id);
                        self.load(conn);
                        return Some(AppMessage::Notify(
                            "Cache entry cleared".to_string(),
                            Severity::Info,
                        ));
                    }
                }
                None
            }
            KeyCode::Char('D') => {
                // Delete all cached files
                for entry in &self.entries {
                    let _ = fs::remove_file(&entry.path);
                }
                let _ = clear_all_cached_paths(conn);
                self.load(conn);
                Some(AppMessage::Notify(
                    "All cache entries cleared".to_string(),
                    Severity::Info,
                ))
            }
            _ => None,
        }
    }
}
